### ISSUE TITLE
FAQ

### ISSUE THREAD

**Issue by waynevanson** [bollian/tree-sitter-openscad#16]
Hey @bollian, I have a few questions I think would come under an FAQ.

- Why did you create a grammar for OpenSCAD?
- Is this grammar complete against the [grammar specification](https://files.openscad.org/grammar.xhtml)? I've had a quick look and it seems good.

I'm personally looking to use this library to create an LSP/formatter for vscode using [Topiary](https://github.com/tweag/topiary) because the current openscad plugins I use have really poor formatting. I assume it's possible with these tools together.

Thank you for putting the grammar together.

**Comment by bollian** [comment]
The original intent behind this parser was to create a language server and to support editor syntax highlighting in tree-sitter-enabled code editors. The first of those has pretty much already happened in the form of https://github.com/Leathong/openscad-LSP, and the second of those has happened to some extent, but not yet to the extent I'm happy with. In the future, I would like to put together the PRs to add this parser to the editors I know about, especially neovim since it's what I use. Stretch goal would be an incremental renderer integrated into a language server, but that's just a pipe dream :).

I actually wasn't aware of that grammar specification you've linked, so thanks for that! And yes, even the names of things in that seem to mostly align with how I've described things in this grammar. While I know there are bugs in this parser (see issue #13), I'm pretty sure we're talking about 1% of cases. It's absolutely the goal to be able to parse 100% of syntactically valid OpenSCAD code in the wild. PRs are welcome and encouraged!

As for your goal of an LSP/formatter, I would really recommend taking a look at that language server I linked earlier. It's certainly a young project, but they've created their own VSCode plugin and I use it with neovim, so it'll be the project that inevitably gets the most attention.
